ansible/roles/vm_set/files/mux_simulator.py

```diff
@@ -116,370 +112,535 @@ def run_cmd(cmdline):
+    '''
 
-    Returns:
-        dict: A dictionary contains full mux status, including connection details, open flow details and active side
-        information. Example:
-            {
-                "bridge": "mbr-vms17-8-0",
-                "vm_set": "vms17-8",
-                "port_index": 0,
-                "ports": {
-                    "nic": "muxy-vms17-8-0",
-                    "upper_tor": "enp59s0f1.3216",
-                    "lower_tor": "enp59s0f1.3272"
-                },
-                "active_side": "upper_tor",
-                "active_port": "enp59s0f1.3216",
-                "flows": {
-                    "muxy-vms17-8-0":
-                    [
-                        {
-                            "action": "output",
-                            "out_port": "enp59s0f1.3216"
-                        },
-                        {
-                            "action": "output",
-                            "out_port": "enp59s0f1.3272"
-                        }
-                    ],
-                    "enp59s0f1.3216":
-                    [
-                        {
-                            "action": "output",
-                            "out_port": "muxy-vms17-8-0"
-                        }
-                    ]
-                }
+    def __init__(self, vm_set, port_index):
+        self.isvalid = True
+        self.lock = threading.Lock()
+
+        self.vm_set = vm_set
+
+        self.port_index = port_index
+        self.bridge = adaptive_name(MUX_BRIDGE_TEMPLATE, vm_set, port_index)
+
+        self._init_ports()
+
+        # If the mux does not have valid ports attahed, it is invalid
+        if not self._get_ports():
+            self.isvalid = False
+            return
+
+        # Initilize the flows configured on the mux bridge
+        self._init_flows()
+        self._get_flows()
+
+        self.flap_counter = 0
+
+    def debug(self, msg):
+        app.logger.debug('bridge={}, {}'.format(self.bridge, msg))
+
+    def info(self, msg):
+        app.logger.info('bridge={}, {}'.format(self.bridge, msg))
+
+    def error(self, msg):
+        app.logger.error('bridge={}, {}'.format(self.bridge, msg))
+
+    def _init_ports(self):
+        self.ports = {
+            NIC: None,
+            UPPER_TOR: None,
+            LOWER_TOR: None
+        }
+        self.sides = {}
+
+    def _get_ports(self):
+        """Use the 'ovs-vsctl list-ports' command to get the ports attached to the mux bridge.
+
+        Example bridge name: 'mbr-vms17-8-0'. In the bridge name, 'vms17-8' is vm_set. '0' is port_index.
+        Example output of 'ovs-ofctl list-ports' command:
+
+        azure@str2-acs-serv-17:~$ sudo ovs-vsctl list-ports mbr-vms17-8-0
+        enp59s0f1.3216
+        enp59s0f1.3272
+        muxy-vms17-8-0
+
+        Returns:
+            boolean: Return False if it is not a valid mux bridge
+        """
+        out = run_cmd('ovs-vsctl list-ports {}'.format(self.bridge))
```

Comment:
How about defind the cmd as a macro, say ```LIST_PORTS_CMD```, just like Line 30~32

Reply:
Good idea!